### PR TITLE
binding: new package for generating netlib bindings

### PR DESCRIPTION
**Pull request by kortschak**
There are no tests for this. Maybe they will come later, but it's not for external consumption and failures will show up in use.

@gonum/developers Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
Failure is unrelated - go1.4.2 needs to go away.

**Review by vladimir-ch — APPROVED** [review]
LGTM

File: binding/binding.go [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+	{Kind: cc.Bool}: template.Must(template.New("bool").Parse("C.bool({{.}})")),
+
+	{Kind: cc.FloatComplex}:                   template.Must(template.New("floatcomplex").Parse("unsafe.Pointer({{.}})")),
+	{Kind: cc.DoubleComplex}:                  template.Must(template.New("doublecomplex").Parse("unsafe.Pointer({{.}})")),
+	{Kind: cc.FloatComplex, IsPointer: true}:  template.Must(template.New("floatcomplex*").Parse("unsafe.Pointer(&{{.}}[0])")),
+	{Kind: cc.DoubleComplex, IsPointer: true}: template.Must(template.New("doublecomplex*").Parse("unsafe.Pointer(&{{.}}[0])")),
+}
+
+// GoTypeFor returns a string representation of the given type using a mapping in
```
Comment (sbinet) [inline binding/binding.go:157]: `s/GoTypeFor/CgoConversionFor/`

File: binding/binding.go [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+		if err != nil {
+			panic(err)
+		}
+		return buf.String()
+	}
+	panic(fmt.Sprintf("unknown type key: %+v", TypeKey{Kind: k, IsPointer: isPtr}))
+}
+
+// CgoConversionForEnum returns a string representation of the given enum type using a mapping
+// in types. GoTypeForEnum will panic if no type mapping is found after searching the
```
Comment (sbinet) [inline binding/binding.go:203]: `s/GoTypeForEnum/CgoConversionForEnum/`

File: binding/binding.go [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+	{Kind: cc.Bool}: template.Must(template.New("bool").Parse("C.bool({{.}})")),
+
+	{Kind: cc.FloatComplex}:                   template.Must(template.New("floatcomplex").Parse("unsafe.Pointer({{.}})")),
+	{Kind: cc.DoubleComplex}:                  template.Must(template.New("doublecomplex").Parse("unsafe.Pointer({{.}})")),
+	{Kind: cc.FloatComplex, IsPointer: true}:  template.Must(template.New("floatcomplex*").Parse("unsafe.Pointer(&{{.}}[0])")),
+	{Kind: cc.DoubleComplex, IsPointer: true}: template.Must(template.New("doublecomplex*").Parse("unsafe.Pointer(&{{.}}[0])")),
+}
+
+// GoTypeFor returns a string representation of the given type using a mapping in
+// types. GoTypeFor will panic if no type mapping is found after searching the
```
Comment (sbinet) [inline binding/binding.go:158]: `s/GoTypeFor/CgoConversionFor/`

File: binding/binding.go [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+				}
+				doc[fn.Name.String()] = fn.Doc.List
+			}
+		}
+	}
+
+	return docs, nil
+}
+
+// Declaration is a description os a C function declaration.
```
Comment (sbinet) [inline binding/binding.go:286]:
``` diff
-os a C function
+of a C function
```

File: binding/binding.go [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+unsigned __builtin_bswap32 (unsigned x);
+unsigned long long __builtin_bswap64 (unsigned long long x);
+`,
+		paths,
+		model(),
+		cc.IncludePaths(includePaths),
+		cc.SysIncludePaths(sysIncludePaths),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse %q: %v", paths, err)
```
Comment (sbinet) [inline binding/binding.go:346]: prepend `binding:` to the error string ?

File: binding/binding.go [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+		p[i] = Parameter{c}
+	}
+	return p
+}
+
+// Declarations returns the C function declarations in the givel set of file paths.
+func Declarations(paths ...string) ([]Declaration, error) {
+	predefined, includePaths, sysIncludePaths, err := cc.HostConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get host config: %v", err)
```
Comment (sbinet) [inline binding/binding.go:327]: prepend `binding:` to the error string ?

**Comment by sbinet** [comment]
apologies for being a tad late to the party

**Comment by kortschak** [comment]
Would you send an issue for me to fix those?

**Comment by sbinet** [comment]
done: https://github.com/gonum/internal/pull/37

**Pull request merged**